src/test/java/frc/robot/util/FieldElementTest.java

```diff
@@ -0,0 +1,364 @@
+    public void testRedNodeRotation() {
+        // The scorePose should have the correct rotation
+
+        Node n = new Node((Vision)null, Alliance.Red, 2, 5);
+        double ang = n.scorePose.getRotation().getRadians();
+        ang = Math.IEEEremainder(ang, 2 * Math.PI);
+        assertEquals(0.0, ang, 0.000001);
+    }
+
+    /**
+     * Transform a Blue Alliance pose to a Red Alliance pose.
+     * 
+     * <p> This should be a library method.
+     * 
+     * <p> Translate and negate x-coordinate.
+     * <p> Copy y-coordinate.
+     * <p> mirror rotation.
+     * 
+     * <p> The Pose2d object does not have a mirror operation.
+     * It is just a 2-d position and a rotation.
+     * @param pose the Blue Alliance pose
+     * @return the Red Alliance pose
+     */
+    public Pose2d mirrorPoseAboutFieldCenter(Pose2d pose) {
+        // mirror the x-coordinate about the center of the field
+        double x = widthField - pose.getX();
+        // the y-coordinate stays the same
+        double y = pose.getY();
+        // The angle from positive x-axis is changed to be minus from negative x-axis
+        double ang = pose.getRotation().getRadians();
+        Rotation2d r = new Rotation2d(Math.PI - ang);
+
+        return new Pose2d(x, y, r);
+    }
+
+    /**
+     * Possibly mirror a Blue Alliance pose
+     * @param pose
+     * @param alliance
+     * @return
+     */
+    public Pose2d mirrorPoseAlliance(Pose2d poseBlue, Alliance alliance) {
+        switch (alliance) {
+            default:
+            case Blue: return poseBlue;
+            case Red: return mirrorPoseAboutFieldCenter(poseBlue);
+        }
+    }
+
+    /**
+     * Test poses for numerical equivalence.
+     * @param p1 first pose
+     * @param p2 second pose
+     * @param eps tolerance for floating point comparison
+     * @returns true if the poses are close enough
+     */
+    public boolean testEquals(Pose2d p1, Pose2d p2, double eps) {
+        return Math.abs(p1.getX() - p2.getX()) < eps &&
+            Math.abs(p1.getY() - p2.getY()) < eps &&
+            Math.abs(Math.IEEEremainder(p1.getRotation().getRadians() - p2.getRotation().getRadians(), 2 * Math.PI)) < eps;
+    }
+
+    /**
+     * Basic tests of field mirroring.
+     */
+    @Test
+    public void testMirror() {
+        // make a test pose
+        Pose2d pose = new Pose2d(2.1, 3.6, new Rotation2d(Math.PI/4));
+        // mirror that pose
+        Pose2d poseCheck = mirrorPoseAboutFieldCenter(pose);
+        // calculate the expected pose
+        Pose2d poseExpect = new Pose2d(widthField - 2.1, 3.6, new Rotation2d(0.75 * Math.PI));
+ 
+        // broken out tests
+        assertEquals(poseExpect.getX(), poseCheck.getX(), 0.001);
+        assertEquals(poseExpect.getY(), poseCheck.getY(), 0.001);
+        assertEquals(poseExpect.getRotation().getRadians(), poseCheck.getRotation().getRadians(), 0.001);
+
+        // integrated test
+        assertTrue(testEquals(poseCheck, poseExpect, 0.001));
```

Comment:
is there a reason to do both integrated and seperate tests?

Reply:
Writing tests of unit tests would be beyond redundant.
As for the point about unit tests, I found they increase my builds by ~2-4 seconds.
Lets be safe and say 0.1 second.
0.1 billion seconds = 3.1 years.
Of course, that's all of the unit tests together currently on main, and it's probably some initialization that takes that time.
I don't really see a justification for or against the test.